source/fab/builder.py

```diff
@@ -119,43 +109,67 @@ def __init__(self,
             workspace.mkdir(parents=True)
 
         self._state = SqliteStateDatabase(workspace)
-        self._target = target
-        self._exec_name = exec_name
-
-        self._command_flags_map: Dict[Type[Command], List[str]] = {}
-        if fpp_flags != '':
-            self._command_flags_map[FortranPreProcessor] = (
-                fpp_flags.split()
-            )
-        if fc_flags != '':
-            self._command_flags_map[FortranCompiler] = (
-                fc_flags.split()
-            )
-        if ld_flags != '':
-            self._command_flags_map[FortranLinker] = (
-                ld_flags.split()
-            )
-        self._queue = QueueManager(n_procs - 1)
 
-    def _extend_task_queue(self, task: Task) -> None:
-        self._queue.add_to_queue(task)
-        for prereq in task.prerequisites:
-            self._queue.add_to_queue(HashCalculator(FileTextReader(prereq),
-                                                    self._state))
+        # If not given, use the name of the target as the exec name
+        # TODO: Should this be in the CLI part?
+        if exec_name == '':
+            exec_name = target
+
+        # Path maps tell the engine what filetype and starting state
+        # the Artifacts representing any files encountered by the
+        # initial descent should have
+        path_maps = [
+            PathMap(r'.*\.f90', FortranSource, Raw),
+            PathMap(r'.*\.F90', FortranSource, Seen),
+        ]
+
+        # Initialise the required Tasks, providing them with any static
+        # properties such as flags to use, workspace location etc
+        # TODO: Eventually the tasks may instead access many of these
+        # properties via the configuration (at Task runtime, to allow for
+        # file-specific overrides?)
+        fortran_preprocessor = FortranPreProcessor(
+            'cpp', ['-traditional-cpp', '-P'] + fpp_flags.split(), workspace
+        )
+        fortran_analyser = FortranAnalyser(workspace)
+        fortran_compiler = FortranCompiler(
+            'gfortran',
+            ['-c', '-J', str(workspace)] + fc_flags.split(), workspace
+        )
+        # TODO: Linker class is probably generic not Fortran specific
+        # now that we are passing in the command to use?
```

Comment:
It always was but that was not important when we were only considering Fortran. Not something for this changeset.

Reply:
It seems like something which will make sense to do when we add C source (rather than writing a CLinker class that is almost identical, and because hopefully the linking stage is designed purely around linking binary objects and won't care or know what the language underneath is)